core/iwasm/aot/aot_runtime.c

```diff
@@ -4104,6 +4105,48 @@ aot_frame_update_profile_info(WASMExecEnv *exec_env, bool alloc_frame)
 #endif /* end of WASM_ENABLE_AOT_STACK_FRAME != 0 */
 
 #if WASM_ENABLE_DUMP_CALL_STACK != 0
+void 
+aot_iterate_callstack(WASMExecEnv *exec_env, const wasm_frame_callback frame_handler, void* user_data)
+{
+/*
+* Note for devs: please refrain from such modifications inside of aot_iterate_callstack
+ * - any allocations/freeing memory
+ * - dereferencing any pointers other than: exec_env, exec_env->module_inst,
+ * exec_env->module_inst->module, pointers between stack's bottom and top_boundary
+ * For more details check wasm_iterate_callstack in wasm_export.h
+*/
+    if (!is_tiny_frame(exec_env)) {
+        //TODO: support standard frames
+        return;
+    }
+    uint8* top_boundary = exec_env->wasm_stack.top_boundary;
+    uint8* top = exec_env->wasm_stack.top;
+    uint8* bottom = exec_env->wasm_stack.bottom;
+
+    bool is_top_index_in_range = top_boundary >= top && top >= (bottom + sizeof(AOTTinyFrame));
+    if (!is_top_index_in_range) {
+        return;
+    }
+    bool is_top_aligned_with_bottom = (unsigned long)(top - bottom) % sizeof(AOTTinyFrame) == 0;
+    if (!is_top_aligned_with_bottom) {
+        return;
+    }
+
+    AOTTinyFrame* frame = (AOTTinyFrame*)(top - sizeof(AOTTinyFrame));
```

Comment:
I wonder if we want to iterate from top to bottom or from bottom to top? bottom pointer is constant but the top one might change during the signal; so if you iterate from top, you might actually start with some very high address if you're unluckly because the pointer will be invalid.

Reply:
> it's better to handle cases where the stack is big and it's better to process only top 100 frames for example

I don't have a hard data on that but I guess in "average application" stack traces aren't expected to be that big? Or can we make the `100` configurable? 

> for wasm it's much harder to go bottom to top because there are no pointers that point to next frame and I haven't found a way to calculate start of the next frame on the stack manually

Agree; however, AOTTinyFrame has a fixed size, so navigating top/bottom is just by moving pointer by sizeof(AOTTinyFrame) bytes.

Reply:
Also before that I do a check that `top` is aligned with bottom, so even it can be invalid it does the same thing as if we started from the bottom and can be controlled through the callback the same way